c_src/membrane_audio_mix_plugin/mixer.c

```diff
@@ -1,210 +1,246 @@
-    }
+  for (uint32_t i = 0; i < state->queue_length; ++i) {
+    state->queue[i] = (int64_t)(state->queue[i] * quotient);
+  }
 
-    for (unsigned int i = 0; i < values_length; ++i)
-    {
-        values[i] = (int64_t)(values[i] * quotient);
-    }
+  for (uint32_t i = 0; i < values_length; ++i) {
+    values[i] = (int64_t)(values[i] * quotient);
+  }
 }
 
-void get_samples(unsigned char *samples, int64_t *values, unsigned int values_length, UnifexState *state)
-{
-    int64_t min = state->sample_max;
-    int64_t max = state->sample_min;
-
-    for (unsigned int i = 0; i < state->queue_length; ++i)
-    {
-        if (state->queue[i] < min)
-        {
-            min = state->queue[i];
-        }
-
-        if (state->queue[i] > max)
-        {
-            max = state->queue[i];
-        }
+/**
+ * Takes given values and values in queue and coverts them to samples.
+ *
+ * If any of the values overflows limits of the format, values will be
+ * scaled down so the peak of the wave will become
+ * maximal (minimal) allowed value.
+ */
+void get_samples(uint8_t *samples, int64_t *values, uint32_t values_length,
+                 UnifexState *state) {
+  int64_t min = state->sample_max;
+  int64_t max = state->sample_min;
+
+  for (uint32_t i = 0; i < state->queue_length; ++i) {
+    if (state->queue[i] < min) {
+      min = state->queue[i];
     }
 
-    for (unsigned int i = 0; i < values_length; ++i)
-    {
-        if (values[i] < min)
-        {
-            min = values[i];
-        }
-
-        if (values[i] > max)
-        {
-            max = values[i];
-        }
-    }
-
-    if (min < state->sample_min)
-    {
-        double quotient = min / (state->sample_min * 1.0);
-        scale(values, values_length, quotient, state);
-    }
-    else if (max > state->sample_max)
-    {
-        double quotient = max / (state->sample_max * 1.0);
-        scale(values, values_length, quotient, state);
+    if (state->queue[i] > max) {
+      max = state->queue[i];
     }
+  }
 
-    int samples_start = 0;
-    for (unsigned int i = 0; i < state->queue_length; ++i, samples_start += state->sample_size)
-    {
-        value_to_sample(state->queue[i], samples + samples_start, state->caps);
+  for (uint32_t i = 0; i < values_length; ++i) {
+    if (values[i] < min) {
+      min = values[i];
     }
 
-    for (unsigned int i = 0; i < values_length; ++i, samples_start += state->sample_size)
-    {
-        value_to_sample(values[i], samples + samples_start, state->caps);
+    if (values[i] > max) {
+      max = values[i];
     }
-
-    unifex_free(state->queue);
-    state->queue = NULL;
-    state->queue_length = 0;
+  }
+
+  if (min < state->sample_min) {
+    double quotient = (state->sample_min * 1.0) / min;
+    scale(values, values_length, quotient, state);
+  } else if (max > state->sample_max) {
+    double quotient = (state->sample_max * 1.0) / max;
+    scale(values, values_length, quotient, state);
+  }
+
+  int32_t samples_size = 0;
```

Comment:
Instead of keeping a number, you could keep a pointer initialized with `samples`